storage-kit/src/traits.rs

```diff
@@ -0,0 +1,128 @@
+use tokio::sync::mpsc;
+
+use crate::{FileChunk, FileKey, FileMetadata};
+
+/// Storage interface to be implemented by the storage providers.
+pub trait Storage: Clone + Send + Sync + 'static {
+    /// Get metadata for a file.
+    fn get_metadata(
+        &self,
+        key: &FileKey,
+    ) -> impl std::future::Future<Output = Option<FileMetadata>> + Send;
+
+    /// Set metadata for a file.
+    fn set_metadata(
+        &self,
+        key: &FileKey,
+        metadata: &FileMetadata,
+    ) -> impl std::future::Future<Output = ()> + Send;
+
+    /// Get a file chunk from storage.
+    fn get_chunk(
+        &self,
+        key: &FileKey,
+        chunk: u64,
+    ) -> impl std::future::Future<Output = Option<FileChunk>> + Send;
+
+    /// Write a file chunk in storage.
+    fn write_chunk(
+        &self,
+        key: &str,
+        chunk: u64,
+        data: &FileChunk,
+    ) -> impl std::future::Future<Output = ()> + Send;
+}
+
+/// The `Actor` trait represents an actor, which runs on its own event loop and can handle messages.
+/// The struct implementing this trait can be seen as the context of the actor, holding the internal
+/// state or the shared data (through commands and queries).
+pub trait Actor: Sized {
+    /// The type of message that the actor can handle.
+    /// Usually an enum that represents the different types of messages that the actor can receive.
+    type Message: Send + Sized + 'static;
+
+    /// The event loop associated with the actor.
+    /// If no custom event loop is needed, the default `EventLoop<Self>` can be used.
+    type EventLoop: ActorEventLoop<Self> + Send + 'static;
+
+    /// Handles a message received by the actor.
+    ///
+    /// * `message` - The message to be handled.
+    ///
+    fn handle_message(
```

Comment:
I'd really like to have the same interface as Actix where you do `Handle<T: Message>` for each _different_ message type but the problem is you end up either having to do dynamic dispatch or have to do very heavy code generation (at least seems that is possible this way). The latter wouldn't work on anything written outside the crate anyway.

Here `Message` is an `enum` that contains all possible commands/queries. For this specific part ("lower-level" modules) I think is the best approach anyway since you don't really get any benefit by isolating implementations for each message. We also get the static analysis and the autocomplete for `P2PCommand::`.